{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This notebook is used to obtain the additional metadata information for the studies served on D2H2. There are files that are read for manual annotation of the studies as well as automatic annotation using the GEO studes supplementary information. The additional metadata includes the tissue types, disease types, platform types, and perturbations for each study. The output is a dictionary that stores the metadata for each study within each species. In addition, a `metadata_counts.json` file is outputted that will be used to popualte the about page with the statisical distributions for the various metadata collected. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import GEOparse\n",
    "import os\n",
    "import json\n",
    "import ftfy\n",
    "from collections import defaultdict \n",
    "import s3fs\n",
    "import os\n",
    "import json\n",
    "import pandas as pd\n",
    "from dotenv import load_dotenv\n",
    "load_dotenv()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Initializing variables for the metadata analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['d2h2/data', 'd2h2/t2d-datasets']\n",
      "145\n",
      "145\n",
      "30\n",
      "17\n"
     ]
    }
   ],
   "source": [
    "\n",
    "endpoint = os.environ.get('ENDPOINT', 'https://d2h2.s3.amazonaws.com/')\n",
    "\n",
    "#USE THE BELOW IF YOU HAVE A KEY AND SECRET VARIABLE\n",
    "base_url = os.environ.get('BASE_URL', 'd2h2/data')\n",
    "s3 = s3fs.S3FileSystem(key = os.environ.get('AWS_ACCESS_KEY_ID'), secret = os.environ.get('AWS_SECRET_ACCESS_KEY'))\n",
    "print(s3.ls('d2h2'))\n",
    "print(len(list(s3.walk(f'{base_url}/mouse', maxdepth=1))[0][1]))\n",
    "print(len(list(s3.walk(f'{base_url}/human', maxdepth=1))[0][1]))\n",
    "print(len(list(s3.walk(f'{base_url}/mouse_single', maxdepth=1))[0][1]))\n",
    "print(len(list(s3.walk(f'{base_url}/human_single', maxdepth=1))[0][1]))\n",
    "dict_storing_gses = {'mouse':list(s3.walk(f'{base_url}/mouse', maxdepth=1))[0][1], 'human': list(s3.walk(f'{base_url}/human', maxdepth=1))[0][1],\n",
    "                      'mouse_single': list(s3.walk(f'{base_url}/mouse_single', maxdepth=1))[0][1],\n",
    "                      'human_single': list(s3.walk(f'{base_url}/human_single', maxdepth=1))[0][1]}\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "dict_storing_information = {'human':{'assay': defaultdict(int), 'disease': defaultdict(int), 'tissue': defaultdict(int), 'platform': defaultdict(int)},\n",
    "                            'mouse':{'assay': defaultdict(int), 'disease': defaultdict(int), 'tissue': defaultdict(int), 'platform': defaultdict(int)}}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "dict_storing_information['human_single'] = {'assay': defaultdict(int), 'disease': defaultdict(int),\n",
    "                                            'tissue': defaultdict(int), 'cell type': defaultdict(int), 'platform': defaultdict(int)}\n",
    "dict_storing_information['mouse_single'] = {'assay': defaultdict(int), 'disease': defaultdict(int),\n",
    "                                            'tissue': defaultdict(int), 'cell type': defaultdict(int), 'platform': defaultdict(int)}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "dict_storing_non_labeled_disease_types = {'human':[], 'mouse': [], 'human_single': [], 'mouse_single': []}\n",
    "dict_storing_non_labeled_tissue_types = {'human':[], 'mouse': [], 'human_single': [], 'mouse_single': []}\n",
    "dict_storing_mislabeled_diabetes = {'human':[], 'mouse': [], 'human_single': [], 'mouse_single': []}\n",
    "dict_storing_all_information = {'assay': defaultdict(int), 'disease': defaultdict(int), 'tissue': defaultdict(int), 'platform': defaultdict(int), 'cell type': defaultdict(int)}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "dict_to_combine_with_metadata = {'human': {}, 'mouse': {}, 'human_single': {}, 'mouse_single': {}}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "list_of_keys_storing_studies = ['human', 'mouse', 'human_single', 'mouse_single']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Manual Disease Labels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "list_storing_manually_annotated_gses = []\n",
    "mouse_bulk_df = pd.read_csv('mouse_nonlabeled_diseases_manually_labelled.csv', index_col=0)\n",
    "list_storing_manually_annotated_gses.extend(mouse_bulk_df['GSE'].to_list())\n",
    "human_bulk_df = pd.read_csv('human_nonlabeled_diseases_manually_labelled.csv', index_col=0)\n",
    "list_storing_manually_annotated_gses.extend(human_bulk_df['GSE'].to_list())\n",
    "list_storing_manually_annotated_gses = [x.strip() for x in list_storing_manually_annotated_gses]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Helper Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "def match_disease_from_manual_annotations(dict_key_for_studies, value):\n",
    "    if 'healthy' in value.lower() and 'gene' in value.lower():\n",
    "        dict_storing_information[dict_key_for_studies]['disease']['Healthy: Gene KO Focus'] +=1\n",
    "        dict_storing_all_information['disease']['Healthy: Gene KO Focus'] +=1\n",
    "        return 'Healthy: Gene KO Focus'\n",
    "    elif 'healthy' in value.lower() and 'treatment' in value.lower():\n",
    "        dict_storing_information[dict_key_for_studies]['disease']['Healthy: Treatment Focus'] +=1\n",
    "        dict_storing_all_information['disease']['Healthy: Treatment Focus'] +=1\n",
    "        return 'Healthy: Treatment Focus'\n",
    "    elif 'healthy' in value.lower() and 'diet' in value.lower():\n",
    "        dict_storing_information[dict_key_for_studies]['disease']['Healthy: Diet Focus'] +=1\n",
    "        dict_storing_all_information['disease']['Healthy: Diet Focus'] +=1\n",
    "        return 'Healthy: Diet Focus'\n",
    "    elif 'healthy' in value.lower() and 'age' in value.lower():\n",
    "        dict_storing_information[dict_key_for_studies]['disease']['Healthy: Age Focus'] +=1\n",
    "        dict_storing_all_information['disease']['Healthy: Age Focus'] +=1\n",
    "        return 'Healthy: Age Focus'\n",
    "    elif 'healthy' in value.lower():\n",
    "        dict_storing_information[dict_key_for_studies]['disease']['Healthy: No Specific Disease Mention'] +=1\n",
    "        dict_storing_all_information['disease']['Healthy: No Specific Disease Mention'] +=1\n",
    "        return 'Healthy: No Specific Disease Mention'\n",
    "    elif 'nash' in value.lower():\n",
    "        dict_storing_information[dict_key_for_studies]['disease']['Nonalcoholic Steatohepatitis (NASH)'] +=1\n",
    "        dict_storing_all_information['disease']['Nonalcoholic Steatohepatitis (NASH)'] +=1\n",
    "        return 'Nonalcoholic Steatohepatitis (NASH)'\n",
    "    elif 'nafld' in value.lower():\n",
    "        dict_storing_information[dict_key_for_studies]['disease']['Non-alcoholic fatty liver disease (NALFD)'] +=1\n",
    "        dict_storing_all_information['disease']['Non-alcoholic fatty liver disease (NALFD)'] +=1\n",
    "        return 'Non-alcoholic fatty liver disease (NALFD)'\n",
    "    elif 'steatosis' in value.lower():\n",
    "        dict_storing_information[dict_key_for_studies]['disease']['Liver Steatosis'] +=1\n",
    "        dict_storing_all_information['disease']['Liver Steatosis'] +=1\n",
    "        return 'Liver Steatosis'\n",
    "    elif 'obesity' in value.lower():\n",
    "        dict_storing_information[dict_key_for_studies]['disease']['Obesity'] +=1\n",
    "        dict_storing_all_information['disease']['Obesity'] +=1\n",
    "        return 'Obesity'\n",
    "    elif 'diabetes' in value.lower():\n",
    "        dict_storing_information[dict_key_for_studies]['disease']['Diabetes (Non-specific)'] +=1\n",
    "        dict_storing_all_information['disease']['Diabetes (Non-specific)'] +=1\n",
    "        return 'Diabetes (Non-specific)'\n",
    "    else:\n",
    "        dict_storing_information[dict_key_for_studies]['disease'][value] +=1\n",
    "        dict_storing_all_information['disease'][value] +=1\n",
    "        return value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_clean_tissue(tissue):\n",
    "    if 'islet' in tissue.lower() or 'pancrea' in tissue.lower() or 'beta cell' in tissue.lower():\n",
    "        return 'Pancreas'\n",
    "    elif 'fat' in tissue.lower() or 'adipose' in tissue.lower() or 'adipocyt' in tissue.lower() or 'bat' in tissue.lower():\n",
    "        return 'Adipose Tissue'\n",
    "    elif 'breast' in tissue.lower() or 'mammary gland' in tissue.lower():\n",
    "        return 'Breast'\n",
    "    elif 'liver' in tissue.lower() or 'hepatoc' in tissue.lower():\n",
    "        return 'Liver'\n",
    "    elif 'kidney' in tissue.lower() or 'pancreatic' in tissue.lower():\n",
    "        return 'Kidney'\n",
    "    elif 'skin' in tissue.lower():\n",
    "        return 'Skin'\n",
    "    elif 'brain' in tissue.lower() or 'cortex' in tissue.lower() or 'hypothalam' in tissue.lower() or 'hippocampus' in tissue.lower():\n",
    "        return 'Brain'\n",
    "    elif 'heart' in tissue.lower() or 'cardiac' in tissue.lower() or 'myocard' in tissue.lower():\n",
    "        return 'Heart'\n",
    "    elif 'pbmc' in tissue.lower():\n",
    "        return 'PBMC'\n",
    "    elif 'skeletal' in tissue.lower():\n",
    "        return 'Skeletal Muscle'\n",
    "    elif 'vastus lateral' in tissue.lower() or 'muscle' in tissue.lower() or 'quadr' in tissue.lower():\n",
    "        return 'Muscle'\n",
    "    elif 'blood' in tissue.lower():\n",
    "        return 'Blood'\n",
    "    elif 'bone' in tissue.lower():\n",
    "        return 'Bone'\n",
    "    elif 'stem cell' in tissue.lower():\n",
    "        return 'Blood'\n",
    "    elif 'small intestin' in tissue.lower() or 'ileum' in tissue.lower() or 'small in' in tissue.lower() or 'duodenum' in tissue.lower():\n",
    "        return 'Small Intestine'\n",
    "    elif 'foxo' in tissue.lower():\n",
    "        return 'Pancreas'\n",
    "    elif 'rnv' in tissue.lower() or 'retina' in tissue.lower():\n",
    "        return 'Retina'\n",
    "    elif 'vascular' in tissue.lower():\n",
    "        return 'Blood'\n",
    "    elif 'thyroid' in tissue.lower():\n",
    "        return 'Thyroid'\n",
    "    elif 'esophag' in tissue.lower():\n",
    "        return 'Esophagus'\n",
    "    elif 'lung' in tissue.lower():\n",
    "        return 'Lung'\n",
    "    elif 'bladder' in tissue.lower():\n",
    "        return 'Bladder'\n",
    "    else:\n",
    "        return tissue"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_bto_identifier(tissue):\n",
    "    bto_df = pd.read_csv('bto_identification.csv', index_col=0)\n",
    "    bto_row = bto_df.loc[bto_df.index == tissue]\n",
    "    if len(bto_row) > 0:\n",
    "        tissue_identifier = tissue + ' ' + bto_row.iloc[0]['Identifier']\n",
    "        return tissue_identifier\n",
    "    else:\n",
    "        return tissue\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_disease_identifier(key):\n",
    "    diseases_df = pd.read_csv('diseases.tsv', sep='\\t', index_col=0)\n",
    "    key_to_search = key\n",
    "    if 'Covid' in key:\n",
    "        key_to_search = 'covid-19'\n",
    "    if 'Diabetes' in key and 'Non-specific' in key:\n",
    "        key_to_search = 'diabetes'\n",
    "    if 'Breast Cancer' in key:\n",
    "        key_to_search = 'breast cancer'\n",
    "    if 'Thyroid Cancer' in key:\n",
    "        key_to_search = 'parathyroid carcinoma'\n",
    "    if 'Liver Disease' in key:\n",
    "        key_to_search = 'liver disease'\n",
    "    if 'Non-alcoholic fatty liver disease' in key:\n",
    "        key_to_search = 'non-alcoholic fatty liver disease'\n",
    "    if 'Liver Cancer' in key:\n",
    "        key_to_search = 'hepatocellular carcinoma'\n",
    "    if 'Pulmonary Hypertension' in key:\n",
    "        key_to_search =  'pulmonary hypertension'\n",
    "    doid_row = diseases_df.loc[(diseases_df['Disease'].str.contains(key_to_search.lower()))]\n",
    "    if len(doid_row) > 0:\n",
    "        disease_identifier = key + ' ' + doid_row.iloc[0]['DOID']\n",
    "        return disease_identifier\n",
    "    else:\n",
    "        return key\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Full study annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for key in ['human', 'mouse']:\n",
    "    studies = dict_storing_gses[key]\n",
    "    if 'human' in key:\n",
    "        geo_organism = 'Homo sapiens'\n",
    "    if 'mouse' in key:\n",
    "        geo_organism = 'Mus musculus'\n",
    "    path_to_study_folder = './studies/' + key\n",
    "    if not os.path.exists(path_to_study_folder):\n",
    "        os.makedirs(path_to_study_folder)\n",
    "        \n",
    "    dict_key_for_studies = key\n",
    "    for study in studies:\n",
    "        dict_to_combine_with_metadata[dict_key_for_studies][study] = {}     \n",
    "        gse = GEOparse.get_GEO(geo = study, destdir = f'{path_to_study_folder}/{study}', silent=True)\n",
    "        # ASSAY TYPE CLASSIFICATION\n",
    "        if 'type' in gse.metadata.keys():\n",
    "            for assay in gse.metadata['type']:\n",
    "                if 'Expression profiling by high throughput sequencing' in assay or 'Genome binding/occupancy profiling by high throughput sequencing' in assay or 'Non-coding RNA profiling by high throughput sequencing' in assay:\n",
    "                    dict_storing_information[dict_key_for_studies]['assay']['RNA seq'] +=1\n",
    "                    dict_storing_all_information['assay']['RNA seq'] +=1\n",
    "                    dict_to_combine_with_metadata[dict_key_for_studies][study]['assay'] = 'RNA seq'\n",
    "                    break\n",
    "                elif 'Expression profiling by array' in assay:\n",
    "                    dict_storing_information[dict_key_for_studies]['assay']['Microarray'] +=1\n",
    "                    dict_storing_all_information['assay']['Microarray'] +=1\n",
    "                    dict_to_combine_with_metadata[dict_key_for_studies][study]['assay'] = 'Microarray'\n",
    "                    break\n",
    "                else:\n",
    "                    dict_storing_information[dict_key_for_studies]['assay']['RNA seq'] +=1\n",
    "                    dict_storing_all_information['assay']['RNA seq'] +=1\n",
    "                    dict_to_combine_with_metadata[dict_key_for_studies][study]['assay'] = 'RNA seq'\n",
    "                    print(assay)\n",
    "                    break\n",
    "\n",
    "        ## PLATFORM ID CLASSIFICATION\n",
    "        if 'platform_id' in gse.metadata.keys():\n",
    "            for platform in gse.metadata['platform_id']:\n",
    "                dict_storing_information[dict_key_for_studies]['platform'][platform] +=1\n",
    "                dict_storing_all_information['platform'][platform] +=1\n",
    "                dict_to_combine_with_metadata[dict_key_for_studies][study]['platform'] = platform\n",
    "                break\n",
    "        else:\n",
    "            dict_storing_information[dict_key_for_studies]['platform']['No Platform information'] +=1\n",
    "            dict_storing_all_information['platform']['No Platform information'] +=1\n",
    "            dict_to_combine_with_metadata[dict_key_for_studies][study]['platform'] = 'No Platform information'\n",
    "        \n",
    "        # DISEASE CLASSIFICATION\n",
    "        string_to_concatenate = ''\n",
    "        if 'summary' in gse.metadata.keys() and 'title' in gse.metadata.keys():\n",
    "            for title in gse.metadata['title']:\n",
    "                title = ftfy.fix_text(title)\n",
    "                string_to_concatenate += title\n",
    "            for summary in gse.metadata['summary']:\n",
    "                summary = ftfy.fix_text(summary)\n",
    "                string_to_concatenate += ' '+ summary\n",
    "        \n",
    "            string_to_concatenate = string_to_concatenate.lower()\n",
    "\n",
    "            if 't1d' in string_to_concatenate or \"type 1 diabetes\" in string_to_concatenate or 'type-1-diabetes' in string_to_concatenate:\n",
    "                dict_storing_information[dict_key_for_studies]['disease']['Type 1 Diabetes'] += 1\n",
    "                dict_storing_all_information['disease']['Type 1 Diabetes'] += 1\n",
    "                dict_to_combine_with_metadata[dict_key_for_studies][study]['disease'] = 'Type 1 Diabetes'\n",
    "                disease_with_identifier = get_disease_identifier('Type 1 Diabetes')\n",
    "                dict_to_combine_with_metadata[dict_key_for_studies][study]['disease_type_identifier'] = disease_with_identifier\n",
    "            elif 't2d' in string_to_concatenate or 'type 2 diabetes' in string_to_concatenate:\n",
    "                dict_storing_information[dict_key_for_studies]['disease']['Type 2 Diabetes'] += 1\n",
    "                dict_storing_all_information['disease']['Type 2 Diabetes'] += 1\n",
    "                dict_to_combine_with_metadata[dict_key_for_studies][study]['disease'] = 'Type 2 Diabetes'\n",
    "                disease_with_identifier = get_disease_identifier('Type 2 Diabetes')\n",
    "                dict_to_combine_with_metadata[dict_key_for_studies][study]['disease_type_identifier'] = disease_with_identifier\n",
    "            elif 'diabetes' in string_to_concatenate or 'diabetic' in string_to_concatenate:\n",
    "                dict_storing_information[dict_key_for_studies]['disease']['Diabetes (Non-specific)'] += 1\n",
    "                dict_storing_all_information['disease']['Diabetes (Non-specific)'] += 1\n",
    "                dict_to_combine_with_metadata[dict_key_for_studies][study]['disease'] = 'Diabetes (Non-specific)'\n",
    "                disease_with_identifier = get_disease_identifier('Diabetes (Non-specific)')\n",
    "                dict_to_combine_with_metadata[dict_key_for_studies][study]['disease_type_identifier'] = disease_with_identifier\n",
    "            elif 'pancreas' in string_to_concatenate or 'pancreatic' in string_to_concatenate or 'islet' in string_to_concatenate or 'β-cell' in string_to_concatenate or 'β cell' in string_to_concatenate:\n",
    "                dict_storing_information[dict_key_for_studies]['disease']['Diabetes (Non-specific)'] += 1\n",
    "                dict_storing_all_information['disease']['Diabetes (Non-specific)'] += 1\n",
    "                dict_to_combine_with_metadata[dict_key_for_studies][study]['disease'] = 'Diabetes (Non-specific)'\n",
    "                dict_storing_mislabeled_diabetes[dict_key_for_studies].append(study)\n",
    "                disease_with_identifier = get_disease_identifier('Diabetes (Non-specific)')\n",
    "                dict_to_combine_with_metadata[dict_key_for_studies][study]['disease_type_identifier'] = disease_with_identifier\n",
    "            #Using the manually labelled data to fill in the nondiabetes related information\n",
    "            else:\n",
    "                if dict_key_for_studies == 'mouse':\n",
    "                    mouse_bulk_df = pd.read_csv('mouse_nonlabeled_diseases_manually_labelled.csv', index_col=0)\n",
    "                    row_for_study = mouse_bulk_df.loc[mouse_bulk_df['GSE'] == study]\n",
    "                    if len(row_for_study) > 0:\n",
    "                        value = row_for_study.iloc[0]['Manual Label']\n",
    "                        disease_returned = match_disease_from_manual_annotations(dict_key_for_studies, value)\n",
    "                        dict_to_combine_with_metadata[dict_key_for_studies][study]['disease'] = disease_returned\n",
    "                        disease_with_identifier = get_disease_identifier(disease_returned)\n",
    "                        dict_to_combine_with_metadata[dict_key_for_studies][study]['disease_type_identifier'] = disease_with_identifier\n",
    "                elif dict_key_for_studies == 'human':\n",
    "                    human_bulk_df = pd.read_csv('human_nonlabeled_diseases_manually_labelled.csv', index_col=0)\n",
    "                    row_for_study = human_bulk_df.loc[human_bulk_df['GSE'] == study]\n",
    "                    if len(row_for_study) > 0:\n",
    "                        value = row_for_study.iloc[0]['Manual Label']\n",
    "                        disease_returned = match_disease_from_manual_annotations(dict_key_for_studies, value)\n",
    "                        dict_to_combine_with_metadata[dict_key_for_studies][study]['disease'] = disease_returned\n",
    "                        disease_with_identifier = get_disease_identifier(disease_returned)\n",
    "                        dict_to_combine_with_metadata[dict_key_for_studies][study]['disease_type_identifier'] = disease_with_identifier\n",
    "                else:\n",
    "                    dict_storing_information[dict_key_for_studies]['disease']['Non Diabetes Related'] += 1\n",
    "                    dict_storing_all_information['disease']['Non Diabetes Related'] += 1\n",
    "                    dict_storing_non_labeled_disease_types[dict_key_for_studies].append(study)\n",
    "                    dict_to_combine_with_metadata[dict_key_for_studies][study]['disease'] = 'Non Diabetes Related'\n",
    "                    disease_with_identifier = get_disease_identifier('Non Diabetes Related')\n",
    "                    dict_to_combine_with_metadata[dict_key_for_studies][study]['disease_type_identifier'] = disease_with_identifier\n",
    "\n",
    "\n",
    "\n",
    "        # TISSUE AND CELL TYPE CLASSIFICATION   \n",
    "        # Used to determine the tissue type using the manually annotated ones and if not the manual annotated go into the the gse sample metadata\n",
    "        if study.strip() in list_storing_manually_annotated_gses:\n",
    "            if dict_key_for_studies == 'mouse':\n",
    "                mouse_bulk_df = pd.read_csv('mouse_nonlabeled_diseases_manually_labelled.csv', index_col=0)\n",
    "                row_for_study = mouse_bulk_df.loc[mouse_bulk_df['GSE'] == study]\n",
    "                if len(row_for_study) > 0:\n",
    "                    tissue_val = row_for_study.iloc[0]['Tissue Type']\n",
    "                    dict_storing_information[dict_key_for_studies]['tissue'][tissue_val] += 1\n",
    "                    dict_storing_all_information['tissue'][tissue_val] += 1\n",
    "                    cleaned_tissue = get_clean_tissue(tissue_val)\n",
    "                    dict_to_combine_with_metadata[dict_key_for_studies][study]['tissue_type'] = cleaned_tissue\n",
    "                    tissue_with_identifier = get_bto_identifier(cleaned_tissue)\n",
    "                    dict_to_combine_with_metadata[dict_key_for_studies][study]['tissue_type_identifier'] = tissue_with_identifier\n",
    "            elif dict_key_for_studies == 'human':\n",
    "                human_bulk_df = pd.read_csv('human_nonlabeled_diseases_manually_labelled.csv', index_col=0)\n",
    "                row_for_study = human_bulk_df.loc[human_bulk_df['GSE'] == study]\n",
    "                if len(row_for_study) > 0:\n",
    "                    tissue_val = row_for_study.iloc[0]['Tissue Type']\n",
    "                    dict_storing_information[dict_key_for_studies]['tissue'][tissue_val] += 1\n",
    "                    dict_storing_all_information['tissue'][tissue_val] += 1\n",
    "                    cleaned_tissue = get_clean_tissue(tissue_val)\n",
    "                    dict_to_combine_with_metadata[dict_key_for_studies][study]['tissue_type'] = cleaned_tissue\n",
    "                    tissue_with_identifier = get_bto_identifier(cleaned_tissue)\n",
    "                    dict_to_combine_with_metadata[dict_key_for_studies][study]['tissue_type_identifier'] = tissue_with_identifier\n",
    "        else:\n",
    "            samples = gse.metadata['sample_id']\n",
    "            for samp in samples:\n",
    "                conditions =  gse.gsms[samp].metadata['characteristics_ch1']\n",
    "                tissue_var = False\n",
    "                cell_type_var = False\n",
    "                for cond in conditions:\n",
    "                    if ':' in cond:\n",
    "                        key, value = cond.split(\":\")[0].strip(), cond.split(\":\", 1)[1].strip()\n",
    "                        if 'tissue' in key:\n",
    "                            tissue = value.strip().lower()\n",
    "                            dict_storing_information[dict_key_for_studies]['tissue'][tissue] += 1\n",
    "                            dict_storing_all_information['tissue'][tissue] += 1\n",
    "                            tissue_var = True\n",
    "                            cleaned_tissue = get_clean_tissue(tissue)\n",
    "                            dict_to_combine_with_metadata[dict_key_for_studies][study]['tissue_type'] = cleaned_tissue\n",
    "                            tissue_with_identifier = get_bto_identifier(cleaned_tissue)\n",
    "                            dict_to_combine_with_metadata[dict_key_for_studies][study]['tissue_type_identifier'] = tissue_with_identifier\n",
    "                            break\n",
    "                if not tissue_var:\n",
    "                    if dict_key_for_studies == 'human' and study.strip() not in list_storing_manually_annotated_gses:\n",
    "                        human_tissue_df = pd.read_csv('human_nonlabeled_tissues_manually_labelled.csv', index_col=0)\n",
    "                        row_for_study = human_tissue_df.loc[human_tissue_df['GSE'] == study]\n",
    "                        if len(row_for_study) > 0:\n",
    "                            tissue_val = row_for_study.iloc[0]['Tissue Type']\n",
    "                            dict_storing_information[dict_key_for_studies]['tissue'][tissue_val] += 1\n",
    "                            dict_storing_all_information['tissue'][tissue_val] += 1\n",
    "                            cleaned_tissue = get_clean_tissue(tissue_val)\n",
    "                            dict_to_combine_with_metadata[dict_key_for_studies][study]['tissue_type'] = cleaned_tissue\n",
    "                            tissue_with_identifier = get_bto_identifier(cleaned_tissue)\n",
    "                            dict_to_combine_with_metadata[dict_key_for_studies][study]['tissue_type_identifier'] = tissue_with_identifier\n",
    "                    if dict_key_for_studies == 'mouse' and study.strip() not in list_storing_manually_annotated_gses:\n",
    "                        mouse_tissue_df = pd.read_csv('mouse_nonlabeled_tissues_manually_labelled.csv', index_col=0)\n",
    "                        row_for_study = mouse_tissue_df.loc[mouse_tissue_df['GSE'] == study]\n",
    "                        if len(row_for_study) > 0:\n",
    "                            tissue_val = row_for_study.iloc[0]['Tissue Type']\n",
    "                            dict_storing_information[dict_key_for_studies]['tissue'][tissue_val] += 1\n",
    "                            dict_storing_all_information['tissue'][tissue_val] += 1\n",
    "                            cleaned_tissue = get_clean_tissue(tissue_val)\n",
    "                            dict_to_combine_with_metadata[dict_key_for_studies][study]['tissue_type'] = cleaned_tissue\n",
    "                            tissue_with_identifier = get_bto_identifier(cleaned_tissue)\n",
    "                            dict_to_combine_with_metadata[dict_key_for_studies][study]['tissue_type_identifier'] = tissue_with_identifier\n",
    "\n",
    "                break\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for key in ['human_single', 'mouse_single']:\n",
    "    single_species_full_info_df = pd.read_csv('{}_all_studies_fully_labelled.csv'.format(key), index_col=0)\n",
    "\n",
    "    studies = dict_storing_gses[key]\n",
    "    if 'human' in key:\n",
    "        geo_organism = 'Homo sapiens'\n",
    "    if 'mouse' in key:\n",
    "        geo_organism = 'Mus musculus'\n",
    "    path_to_study_folder = './studies/' + key\n",
    "    if not os.path.exists(path_to_study_folder):\n",
    "        os.makedirs(path_to_study_folder)\n",
    "    dict_key_for_studies = key\n",
    "    for study in studies:\n",
    "        dict_to_combine_with_metadata[dict_key_for_studies][study] = {}     \n",
    "        gse = GEOparse.get_GEO(geo = study, destdir = f'{path_to_study_folder}/{study}', silent=True)\n",
    "        # ASSAY TYPE CLASSIFICATION\n",
    "        if 'type' in gse.metadata.keys():\n",
    "            for assay in gse.metadata['type']:\n",
    "                if 'Expression profiling by high throughput sequencing' in assay or 'Genome binding/occupancy profiling by high throughput sequencing' in assay or 'Non-coding RNA profiling by high throughput sequencing' in assay:\n",
    "                    dict_storing_information[dict_key_for_studies]['assay']['RNA seq'] +=1\n",
    "                    dict_storing_all_information['assay']['RNA seq'] +=1\n",
    "                    dict_to_combine_with_metadata[dict_key_for_studies][study]['assay'] = 'RNA seq'\n",
    "                    break\n",
    "                elif 'Expression profiling by array' in assay:\n",
    "                    dict_storing_information[dict_key_for_studies]['assay']['Microarray'] +=1\n",
    "                    dict_storing_all_information['assay']['Microarray'] +=1\n",
    "                    dict_to_combine_with_metadata[dict_key_for_studies][study]['assay'] = 'Microarray'\n",
    "                    break\n",
    "                else:\n",
    "                    dict_storing_information[dict_key_for_studies]['assay']['RNA seq'] +=1\n",
    "                    dict_storing_all_information['assay']['RNA seq'] +=1\n",
    "                    dict_to_combine_with_metadata[dict_key_for_studies][study]['assay'] = 'RNA seq'\n",
    "                    print(assay)\n",
    "                    break\n",
    "\n",
    "        ## PLATFORM ID CLASSIFICATION\n",
    "        if 'platform_id' in gse.metadata.keys():\n",
    "            for platform in gse.metadata['platform_id']:\n",
    "                dict_storing_information[dict_key_for_studies]['platform'][platform] +=1\n",
    "                dict_storing_all_information['platform'][platform] +=1\n",
    "                dict_to_combine_with_metadata[dict_key_for_studies][study]['platform'] = platform\n",
    "                break\n",
    "        else:\n",
    "            dict_storing_information[dict_key_for_studies]['platform']['No Platform information'] +=1\n",
    "            dict_storing_all_information['platform']['No Platform information'] +=1\n",
    "            dict_to_combine_with_metadata[dict_key_for_studies][study]['platform'] = 'No Platform information'\n",
    "        \n",
    "        # DISEASE and Tissue CLASSIFICATION\n",
    "        row_for_study = single_species_full_info_df.loc[single_species_full_info_df['GSE'] == study]\n",
    "        if len(row_for_study) > 0:\n",
    "            disease = row_for_study.iloc[0]['Disease Label']\n",
    "            tissue = row_for_study.iloc[0]['Tissue']\n",
    "            cell_type = row_for_study.iloc[0]['Cell Type']\n",
    "            if 'diabetes' in disease.lower():\n",
    "                dict_storing_information[dict_key_for_studies]['disease']['Diabetes (Non-specific)'] +=1\n",
    "                dict_storing_all_information['disease']['Diabetes (Non-specific)'] +=1\n",
    "                dict_to_combine_with_metadata[dict_key_for_studies][study]['disease'] = 'Diabetes (Non-specific)'\n",
    "                disease_with_identifier = get_disease_identifier('Diabetes (Non-specific)')\n",
    "                dict_to_combine_with_metadata[dict_key_for_studies][study]['disease_type_identifier'] = disease_with_identifier\n",
    "            elif 'healthy' in disease.lower():\n",
    "                dict_storing_information[dict_key_for_studies]['disease']['Healthy: No Specific Disease Mention'] +=1\n",
    "                dict_storing_all_information['disease']['Healthy: No Specific Disease Mention'] +=1\n",
    "                dict_to_combine_with_metadata[dict_key_for_studies][study]['disease'] = 'Healthy: No Specific Disease Mention'\n",
    "                disease_with_identifier = get_disease_identifier('Healthy: No Specific Disease Mention')\n",
    "                dict_to_combine_with_metadata[dict_key_for_studies][study]['disease_type_identifier'] = disease_with_identifier\n",
    "            else:\n",
    "                dict_storing_information[dict_key_for_studies]['disease'][disease] += 1\n",
    "                dict_storing_all_information['disease'][disease] += 1\n",
    "                dict_to_combine_with_metadata[dict_key_for_studies][study]['disease'] = disease\n",
    "                disease_with_identifier = get_disease_identifier(disease)\n",
    "                dict_to_combine_with_metadata[dict_key_for_studies][study]['disease_type_identifier'] = disease_with_identifier\n",
    "\n",
    "            # Tissue type\n",
    "            dict_storing_information[dict_key_for_studies]['tissue'][tissue] += 1\n",
    "            dict_storing_all_information['tissue'][tissue] += 1\n",
    "            cleaned_tissue = get_clean_tissue(tissue)\n",
    "            dict_to_combine_with_metadata[dict_key_for_studies][study]['tissue_type'] = cleaned_tissue\n",
    "            tissue_with_identifier = get_bto_identifier(cleaned_tissue)\n",
    "            dict_to_combine_with_metadata[dict_key_for_studies][study]['tissue_type_identifier'] = tissue_with_identifier\n",
    "\n",
    "            # Cell Type\n",
    "            dict_storing_information[dict_key_for_studies]['cell type'][cell_type] += 1\n",
    "            dict_storing_all_information['cell type'][cell_type] += 1\n",
    "            dict_to_combine_with_metadata[dict_key_for_studies][study]['cell type'] = cell_type\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Count Confirmation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total number of each study type\n",
      "145\n",
      "145\n",
      "17\n",
      "30\n",
      "Number of Assay Measurements\n",
      "145\n",
      "17\n",
      "145\n",
      "30\n",
      "Number of Platform Measurements\n",
      "145\n",
      "17\n",
      "145\n",
      "30\n",
      "Number of Disease Measurements\n",
      "145\n",
      "17\n",
      "145\n",
      "30\n",
      "Total number of tissue types determined\n",
      "145\n",
      "17\n",
      "145\n",
      "30\n",
      "Total number of tissue type classifications determined\n",
      "83\n",
      "58\n",
      "7\n",
      "12\n"
     ]
    }
   ],
   "source": [
    "print('Total number of each study type')\n",
    "print(len(dict_storing_gses['human']))\n",
    "print(len(dict_storing_gses['mouse']))\n",
    "print(len(dict_storing_gses['human_single']))\n",
    "print(len(dict_storing_gses['mouse_single']))\n",
    "print('Number of Assay Measurements')\n",
    "print(sum(map(lambda x : x,dict_storing_information['human']['assay'].values())))\n",
    "print(sum(map(lambda x : x,dict_storing_information['human_single']['assay'].values())))\n",
    "print(sum(map(lambda x : x,dict_storing_information['mouse']['assay'].values())))\n",
    "print(sum(map(lambda x : x,dict_storing_information['mouse_single']['assay'].values())))\n",
    "print('Number of Platform Measurements')\n",
    "print(sum(map(lambda x : x,dict_storing_information['human']['platform'].values())))\n",
    "print(sum(map(lambda x : x,dict_storing_information['human_single']['platform'].values())))\n",
    "print(sum(map(lambda x : x,dict_storing_information['mouse']['platform'].values())))\n",
    "print(sum(map(lambda x : x,dict_storing_information['mouse_single']['platform'].values())))\n",
    "print('Number of Disease Measurements')\n",
    "print(sum(map(lambda x : x,dict_storing_information['human']['disease'].values())))\n",
    "print(sum(map(lambda x : x,dict_storing_information['human_single']['disease'].values())))\n",
    "print(sum(map(lambda x : x,dict_storing_information['mouse']['disease'].values())))\n",
    "print(sum(map(lambda x : x,dict_storing_information['mouse_single']['disease'].values())))\n",
    "print('Total number of tissue types determined')\n",
    "print(sum(map(lambda x : x,dict_storing_information['human']['tissue'].values())))\n",
    "print(sum(map(lambda x : x,dict_storing_information['human_single']['tissue'].values())))\n",
    "print(sum(map(lambda x : x,dict_storing_information['mouse']['tissue'].values())))\n",
    "print(sum(map(lambda x : x,dict_storing_information['mouse_single']['tissue'].values())))\n",
    "print('Total number of tissue type classifications determined')\n",
    "print(len(dict_storing_information['human']['tissue']))\n",
    "print(len(dict_storing_information['mouse']['tissue']))\n",
    "print(len(dict_storing_information['human_single']['tissue']))\n",
    "print(len(dict_storing_information['mouse_single']['tissue']))\n",
    "# print('Total number of consolidated tissue types determined')\n",
    "# print(len(dict_storing_information['human']['tissue_cleaned']))\n",
    "# print(len(dict_storing_information['mouse']['tissue_cleaned']))\n",
    "# print(len(dict_storing_information['human_single']['tissue_cleaned']))\n",
    "# print(len(dict_storing_information['mouse_single']['tissue_cleaned']))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# WHOLE DATASET CLASSIFCATION ANNOTATIONS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Consolidating Certain Terms Manually as of now\n",
    "\n",
    "dict_storing_all_information['tissue_cleaned'] = defaultdict(int)\n",
    "for tissue, count in dict_storing_all_information['tissue'].items():\n",
    "    if 'islet' in tissue.lower() or 'pancrea' in tissue.lower() or 'beta cell' in tissue.lower():\n",
    "        dict_storing_all_information['tissue_cleaned']['Pancreas'] += count\n",
    "    elif 'fat' in tissue.lower() or 'adipose' in tissue.lower() or 'adipocyt' in tissue.lower() or 'bat' in tissue.lower():\n",
    "        dict_storing_all_information['tissue_cleaned']['Adipose Tissue'] += count\n",
    "    elif 'breast' in tissue.lower() or 'mammary gland' in tissue.lower():\n",
    "        dict_storing_all_information['tissue_cleaned']['Breast'] += count\n",
    "    elif 'liver' in tissue.lower() or 'hepatoc' in tissue.lower():\n",
    "        dict_storing_all_information['tissue_cleaned']['Liver'] += count\n",
    "    elif 'kidney' in tissue.lower() or 'pancreatic' in tissue.lower():\n",
    "        dict_storing_all_information['tissue_cleaned']['Kidney'] += count\n",
    "    elif 'skin' in tissue.lower():\n",
    "        dict_storing_all_information['tissue_cleaned']['Skin'] += count\n",
    "    elif 'brain' in tissue.lower() or 'cortex' in tissue.lower() or 'hypothalam' in tissue.lower() or 'hippocampus' in tissue.lower():\n",
    "        dict_storing_all_information['tissue_cleaned']['Brain'] += count\n",
    "    elif 'heart' in tissue.lower() or 'cardiac' in tissue.lower() or 'myocard' in tissue.lower():\n",
    "        dict_storing_all_information['tissue_cleaned']['Heart'] += count\n",
    "    elif 'pbmc' in tissue.lower():\n",
    "        dict_storing_all_information['tissue_cleaned']['PBMC'] += count\n",
    "    elif 'skeletal' in tissue.lower():\n",
    "        dict_storing_all_information['tissue_cleaned']['Skeletal Muscle'] += count\n",
    "    elif 'vastus lateral' in tissue.lower() or 'muscle' in tissue.lower() or 'quadr' in tissue.lower():\n",
    "        dict_storing_all_information['tissue_cleaned']['Muscle'] += count\n",
    "    elif 'blood' in tissue.lower():\n",
    "        dict_storing_all_information['tissue_cleaned']['Blood'] += count\n",
    "    elif 'bone' in tissue.lower():\n",
    "        dict_storing_all_information['tissue_cleaned']['Bone'] += count\n",
    "    elif 'stem cell' in tissue.lower():\n",
    "        dict_storing_all_information['tissue_cleaned']['Blood'] += count\n",
    "    elif 'small intestin' in tissue.lower() or 'ileum' in tissue.lower() or 'small in' in tissue.lower() or 'duodenum' in tissue.lower():\n",
    "        dict_storing_all_information['tissue_cleaned']['Small Intestine'] += count\n",
    "    elif 'foxo' in tissue.lower():\n",
    "        dict_storing_all_information['tissue_cleaned']['Pancreas'] += count\n",
    "    elif 'rnv' in tissue.lower() or 'retina' in tissue.lower():\n",
    "        dict_storing_all_information['tissue_cleaned']['Retina'] += count\n",
    "    elif 'vascular' in tissue.lower():\n",
    "        dict_storing_all_information['tissue_cleaned']['Blood'] += count\n",
    "    elif 'thyroid' in tissue.lower():\n",
    "        dict_storing_all_information['tissue_cleaned']['Thyroid'] += count\n",
    "    elif 'esophag' in tissue.lower():\n",
    "        dict_storing_all_information['tissue_cleaned']['Esophagus'] += count\n",
    "    elif 'lung' in tissue.lower():\n",
    "        dict_storing_all_information['tissue_cleaned']['Lung'] += count\n",
    "    elif 'bladder' in tissue.lower():\n",
    "        dict_storing_all_information['tissue_cleaned']['Bladder'] += count\n",
    "    else:\n",
    "        dict_storing_all_information['tissue_cleaned'][tissue] += count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Sorting the data and creating legend for pie charts\n",
    "legend_dict_full = defaultdict()\n",
    "\n",
    "for metadata_key, dictionary_unsorted in dict_storing_all_information.items():\n",
    "    dictionary_sorted = sorted(dict_storing_all_information[metadata_key].items(), key = lambda x:x[1], reverse=True)\n",
    "    new_dict = {}\n",
    "    for item in dictionary_sorted:\n",
    "        new_dict[item[0]] = item[1]\n",
    "    dict_storing_all_information[metadata_key] = new_dict\n",
    "    \n",
    "    if 'tissue' in metadata_key or 'disease' in metadata_key:\n",
    "        temp_dict = {}\n",
    "        for key, value in dict_storing_all_information[metadata_key].items():\n",
    "            percent = str( round(float(value/347 *100), 2) ) + '%'\n",
    "            temp_dict[key] = key + \" \" + percent\n",
    "        legend_dict_full[metadata_key] = temp_dict"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Output Data Generation files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for species, dictionary_of_gses in dict_to_combine_with_metadata.items():\n",
    "    print(species)\n",
    "    for gse, metadata in dictionary_of_gses.items():\n",
    "        print(gse)\n",
    "        metadata_file = base_url + '/' + species + '/' + gse + '/' + gse + '_Metadata.tsv'\n",
    "        metadata_dataframe = pd.read_csv(s3.open(metadata_file), sep='\\t')\n",
    "        list_of_metadata_perts = metadata_dataframe.iloc[:,1].unique().tolist()\n",
    "        list_of_metadata_perts = list(set([x.strip() for x in list_of_metadata_perts]))\n",
    "        dict_to_combine_with_metadata[species][gse]['perturbations'] = list_of_metadata_perts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dict_to_combine_with_metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "with open('./additional_metadata_d2h2.pickle', 'wb') as f:\n",
    "\tpickle.dump(dict_to_combine_with_metadata, f, protocol=pickle.HIGHEST_PROTOCOL)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "assay\n",
      "2\n",
      "disease\n",
      "29\n",
      "platform\n",
      "29\n",
      "cell type\n",
      "18\n",
      "tissue\n",
      "31\n"
     ]
    }
   ],
   "source": [
    "# Create the counts json for each of the metadata fields that will be used to make the interactve plots on the frontend.\n",
    "dict_for_json = {}\n",
    "dict_for_json['assay'] = {'labels':list(dict_storing_all_information['assay'].keys()), 'counts': list(dict_storing_all_information['assay'].values())}\n",
    "dict_for_json['disease'] = {'labels':list(dict_storing_all_information['disease'].keys()), 'counts': list(dict_storing_all_information['disease'].values())}\n",
    "dict_for_json['platform'] = {'labels':list(dict_storing_all_information['platform'].keys()), 'counts': list(dict_storing_all_information['platform'].values())}\n",
    "dict_for_json['cell type'] = {'labels':list(dict_storing_all_information['cell type'].keys()), 'counts': list(dict_storing_all_information['cell type'].values())}\n",
    "dict_for_json['tissue'] = {'labels':list(dict_storing_all_information['tissue_cleaned'].keys()), 'counts': list(dict_storing_all_information['tissue_cleaned'].values())}\n",
    "for key, dict in dict_for_json.items():\n",
    "    print(key)\n",
    "    print(len(dict_for_json[key]['labels']))\n",
    "\n",
    "with open('metadata_counts.json', 'w') as f:\n",
    "    json.dump(dict_for_json, f)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.17"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
