{% extends 'base.html' %}


{% block body %}
<div class="viewer-title">
    <div class="container viewer-title d-flex align-items-center justify-content-center text-center mt-4">
        <h4>
            <a class="brand px-3 py-3 col-12"
                href="{{ gse_metadata_single[species][geo_accession].get('gse_link', ['#'])[0]|safe }}" target="_blank"
                rel="noopener noreferrer" style="text-decoration: none">{{ geo_accession }}: {{
                gse_metadata_single[species][geo_accession].get('title', ["Title Not Found"])[0]|safe}}
            </a>
        </h4>
    </div>
</div>




<div id="species" hidden>{{ species }}</div>
<div id="singlegse" hidden>{{ geo_accession }}</div>


<!-- Abstract small -->
<div class="row justify-content-center">
    <div class="col-10">

        <!-- Abstract large -->
        <div class="d-block">
            <div class="h4 mt-3 justify-content-center" style="display: flex;">
                scRNA-seq Viewer
                <div class="tooltip ml-2">
                    <img src="../static/img/more_info.png" alt="Info" style="width: 20px;"/>
                    <span class="tooltiptext">
                        Studies were incorporated in a barcode, matrix, feature format. This format can be found on the 
                        <a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/output/matrices" target="_blank" rel="noopener noreferrer">
                        10x Genomics website
                        </a> 
                        for processing of single cell studies to obtain the gene expression matrices. 
                        For each study, the metadata incorporated in GEO were manually curated into profiles and the samples were separated based on applicable groups and conditions. 
                        The expression data from the cells for the samples within the same profile and condition were aggregated into an expression matrix with the cell barcodes having the sample name appended to it to ensure unique cell names.
                    </span>
                </div>
            </div>
            <div class="custom-col p-2 mb-5">
                <h4 class="mb-2 ml-2">Description (from GEO)</h4>
                <p class="text-br-red font-weight-bold ml-2">Submission Date: <span class="font-weight-normal"
                        style="color:#212529">{{
                        (gse_metadata_single[species][geo_accession].get('submission_date')[0][:-5]
                        + ',' + gse_metadata_single[species][geo_accession].get('submission_date')[0][-5:])|safe
                        }}</span>
                </p>
                {% for paragraph in gse_metadata_single[species][geo_accession].get('summary', ['Summary Not Found']) %}
                {% if loop.first %}
                <p class="text-br-red font-weight-bold ml-2">Summary: <span class="font-weight-normal"
                        style="color:#212529">{{ paragraph|safe }}</span></p>
                {% else %}
                <p>{{ paragraph|safe }}</p>
                {% endif %}
                {% endfor %}
                <p class="text-br-red font-weight-bold ml-2">GEO Accession ID: <a class="font-weight-normal"
                        style="color: #0e63c4"
                        href="{{ gse_metadata_single[species][geo_accession].get('gse_link', ['#'])[0]|safe }}"
                        target="_blank" rel="noopener noreferrer">{{
                        gse_metadata_single[species][geo_accession].get('geo_accession', ['GEO Accession ID Not
                        Found'])[0]
                        }}</a></p>
                {% if gse_metadata_single[species][geo_accession].get('pubmed_id', ['No Pubmed ID'])[0]|safe != 'No
                Pubmed
                ID' %}
                <p class="text-br-red font-weight-bold ml-2">PMID: <a class="font-weight-normal" style="color: #0e63c4"
                        href="{{ gse_metadata_single[species][geo_accession].get('pubmed_link', ['#'])[0]|safe }}"
                        target="_blank" rel="noopener noreferrer">{{
                        gse_metadata_single[species][geo_accession].get('pubmed_id', ['No Pubmed ID'])[0]|safe }}</a>
                </p>
                {% else %}
                <p class="text-br-red font-weight-bold ml-2">PMID: <span class="font-weight-normal"
                        style="color:#212529">N/A</span></p>
                {% endif %}
            </div>
        </div>
    </div>
</div>


<div class="row justify-content-center mb-3 mt-2">
    <div class="col-10">
        <div class="custom-col justify-content-center" style="overflow: visible">
            <div class="h4 mt-3 justify-content-center" style="display: flex;">
                Select Condition to compute Visualization and Analysis
                <div class="tooltip ml-2">
                    <img src="../static/img/more_info.png" alt="Info" style="width: 20px;"/>
                    <span class="tooltiptext">
                        Preprocessing and downstream analysis were computed using the <a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-017-1382-0" target="_blank" rel="noopener noreferrer">scanpy</a> Python library and the
                        steps of processing followed the <a href="https://satijalab.org/seurat/articles/pbmc3k_tutorial" target="_blank" rel="noopener noreferrer">Seurat vignette</a>. Cells and genes with no expression or very low expression were 
                        removed from the dataset based on a predefined threshold. The data was then normalized across the expression within the cells and log normalized. The top 2000 highly variable genes were extracted to be used for downstream analysis. 
                        For each of these aggregated data matrices, the clusters were computed using the <a href="https://www.nature.com/articles/s41598-019-41695-z" target="_blank" rel="noopener noreferrer">leiden algorithm</a>. Scanpy was then used to compute the PCA, t-SNE, and UMAPs.
                        The points in the plots are labelled by their appropraiate cell type labels. The cell type labels were computed using the <a href="https://scanpy.readthedocs.io/en/stable/generated/scanpy.tl.rank_genes_groups.html" target="_blank" rel="noopener noreferrer">wilcoxon</a> method as the differential gene expression method.
                        The top 250 genes were then used for enrichment analysis against the <a href="https://academic.oup.com/nar/article/51/D1/D870/6775381" target="_blank" rel="noopener noreferrer">CellMarker</a> library in order to determine the most appropriate cell type label with the lowest p-value.
                    </span>
                </div>
            </div>

            <div class="row justify-items-center justify-content-center mb-3 mt-3 text-center">
                <!-- style="max-width: 800px;word-wrap: break-word;white-space: normal;" -->
                <select name="Select Method" id="methodsingle" class="libpicker" style="max-width: 1100px;word-wrap: break-word;white-space: normal;">
                    {% for cond in study_conditions %}
                    <option value="{{cond}}">{{cond}}</option>
                    {% endfor %}
                </select>
            </div>
            <div id="change-loading" class="m-3"></div>
            <div class="h4 pl-2 mt-4 mb-4 text-center">PCA, tSNE, and UMAP Visualizations Labeled by Cell Types:</div>

            <div id="singleplots-loading" class="m-3"></div>
            <div class="row justify-content-center mt-4 singleplotclass centered">
                <ul class="nav nav-tabs">
                    <li class="nav-item">
                        <a class="nav-link active" data-toggle="tab" data-bs-target="#pca" href="#pca">PCA</a>
                    </li>
                    <li class="nav-item">
                        <a class="nav-link" data-toggle="tab" data-bs-target="#tsne" href="#tsne">tSNE</a>
                    </li>
                    <li class="nav-item">
                        <a class="nav-link" data-toggle="tab" data-bs-target="#umap" href="#umap">UMAP</a>
                    </li>
                </ul>

                <!-- Tab panes -->
                <div class="h4 pl-2 mt-4 mb-4 text-center">
                    <div class="tab-content">
                        <div class="tab-pane container active" id="pca">
                            <div id="pca-plot" class="centered"></div>
                        </div>
                        <div class="tab-pane container fade" id="tsne">
                            <div id="tsne-plot" class="centered"></div>
                        </div>
                        <div class="tab-pane container fade" id="umap">
                            <div id="umap-plot" class="centered"></div>
                        </div>
                    </div>
                </div>
            </div>

            <div class="row justify-content-center mt-4 centered custom-col" id="cellplot">
            </div>




            <div class="container px-4 pb-5" style="min-height: 100%;">

                <!-- Title -->
                <div class="row mt-4">

                    <div class="col-12">
                        <div class="h4 pl-2 mb-4">
                            {% if gse_metadata_single[species][geo_accession].get('type') == ['Expression profiling by
                            array'] %}
                            Microarray Data Explorer
                            {% elif gse_metadata_single[species][geo_accession].get('type') == ['Expression profiling by
                            high throughput
                            sequencing'] %}
                            RNA-Seq Data Explorer
                            {% else %}
                            Gene Expression Data Explorer
                            {% endif %}
                        </div>
                    </div>
                </div>

                <!-- Gene Expression Explorer -->
                <div class="row justify-content-center">

                    <!-- Boxplots -->
                    <div class="col-lg-8 col-xl-9 mb-4 justify-content-center">

                        <!-- Selector and plot -->
                        <div class="custom-col justify-content-center text-center">
                            Enter gene symbol:
                            <select id="gene-select" class="m-4 studysearch"
                                data-url-genes="{{ url_for('genes_api_single', geo_accession=geo_accession, condition=study_conditions[0]) }}"
                                data-geo-acc={{geo_accession}}></select>
                            <div id="boxplotloader" class="m-4"></div>
                            <div id="boxplot" class="plotly-plot chart m-4"
                                data-url-plot="{{ url_for('plot_api_single', geo_accession=geo_accession,condition=study_conditions[0]) }}">
                            </div>
                        </div>



                        <div class="d-none d-lg-block">

                        </div>

                    </div>

                    <!-- Table -->
                    <div class="col-lg-4 col-xl-3 p-2">
                        <p class="m-3" style="font-size: 0.97rem;">Select conditions below to toggle them from the plot:
                        </p>

                        <table id="condition-table" class="text-center">
                            <thead>
                                <tr>

                                    <!-- <th>Clusters</th> -->
                                    <th>Cell Types</th>
                                    <th>Cell Samples</th>
                                </tr>
                            </thead>
                            <tbody id="singlecell">
                                {% for condition in metadata_dict %}

                                <tr>

                                    <td class="p-0 ml-3 border"><button
                                            style="background-color: #ECEFF1; color: black; width: 150px;word-wrap: break-word;white-space: normal;"
                                            class="btn m-0 rounded-0 py-0 condition-btn active plotted "
                                            data-toggle="button" autocomplete="off"
                                            data-group_label="{{ condition }}">{{ condition|safe }} </button>
                                    </td>

                                    <td class="border" style="padding: 0px 18px; width: max-content;">
                                        <button style="width: max-content;"
                                            class="btn-custom btn-group-sm btn-collapse collapsed d-flex align-items-center text-center"
                                            data-toggle="collapse" data-target="#samples-{{ gse_id }}"
                                            aria-expanded="false" aria-controls="samples-{{ gse_id }}">
                                            <div class="text">{{ metadata_dict_samples[condition] }} Cells
                                            </div>
                                        </button>


                                    </td>


                                </tr>
                                {% endfor %}

                            </tbody>
                        </table>
                    </div>
                </div>
            </div>
            <div class="h4 mt-3 justify-content-center" style="display: flex;">
                Differential Gene Expression Analysis
                <div class="tooltip ml-2">
                    <img src="../static/img/more_info.png" alt="Info" style="width: 20px;"/>
                    <span class="tooltiptext">
                        Differential gene expression can be computed for a single cell type labeled group of cells vs the rest. 
                        These include <a href="https://www.sciencedirect.com/topics/medicine-and-dentistry/wilcoxon-signed-ranks-test" target="_blank" rel="noopener noreferrer">wilcoxon</a>,
                        <a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-014-0550-8" target="_blank" rel="noopener noreferrer">DESeq2,</a> or <a href="https://pubmed.ncbi.nlm.nih.gov/24650281/" target="_blank" rel="noopener noreferrer">characteristic direction</a>.
                    </span>
                </div>
            </div>

            <div class="h6 text-center mt-3"> Select Cluster for differential gene expression analysis:</div>
            <div class="row justify-items-center justify-content-center mb-3 mt-3 text-center">
                <select name="Select Method" id="differentialcluster" class="libpicker">
                    {% for cond in metadata_dict %}
                    <option value="{{cond}}">{{cond}}</option>
                    {% endfor %}
                </select>
            </div>
            <div class="h6 text-center mt-3"> Select differential expression analysis method:</div>
            <div class="row justify-items-center justify-content-center mb-3 mt-3 text-center">
                <select name="Select Method" id="method" class="libpicker" style="width: auto;">
                    <option value="wilcoxon">wilcoxon</option>
                    <option value="DESeq2">DESeq2</option>
                    <option value="characteristic_direction">Characteristic Direction</option>
                </select>
            </div>
            <div id="normalization" style="display: none;">
                <div class="h6 text-center mb-2"> Select Normalization Methods:</div>
                <div class="row justify-items-center justify-content-center">
                    <div class="h9 m-2">logCPM normalization:</div>
                    <label class="switch" for="logCPM">
                        <input type="checkbox" id="logCPM" checked>
                        <span class="slider round"></span>
                    </label>
                    <div class="h9 m-2">log normalization:</div>
                    <label class="switch" for="log">
                        <input type="checkbox" id="log">
                        <span class="slider round"></span>
                    </label>
                    <div class="h9 m-2">Z normalization:</div>
                    <label class="switch" for="z">
                        <input type="checkbox" id="z" checked>
                        <span class="slider round"></span>
                    </label>
                    <div class="h9 m-2">Quantile normalization:</div>
                    <label class="switch" for="q">
                        <input type="checkbox" id="q">
                        <span class="slider round"></span>
                    </label>
                </div>
            </div>

            <div class="row justify-content-center">
                <button id="dge-button-single" class="btn btn-primary mx-auto mt-2 mb-3" style="width: 125px"
                    data-url-samp="{{ url_for('samples_api', geo_accession=geo_accession) }}"
                    data-gse-id="{{ gse_metadata_single[species][geo_accession].get('geo_accession', ['No GSE ID'])[0] }}"
                    data-gpl-id="{{ gse_metadata_single[species][geo_accession].get('cur_gpl', ['No GPL ID'])[0] }}">
                    Compute DGE
                </button>
            </div>

            <div id="dge-loading" class="m-3"></div>
            <div id="dge-plot" class="centered"></div>
            <div id="geneset-buttons" class="justify-content-center"></div>
            <div class="col-10 justify-content-center mx-auto" id="dge-table-area"></div>
            <div id="enrichment-area" class="justify-content-center"></div>

            <div id="dgea-loading" class="m-3"></div>
        </div>
    </div>
</div>


{% endblock %}

{% block footer %}
<script>
    const el = document.getElementById('method');

    const box = document.getElementById('normalization');

    $("#singleviewer").addClass('active');
    $("#singleviewer").parents('li').addClass('active');


    generate_single_plots()


</script>



{% endblock %}