{% extends 'base.html' %}

{% block html %}class='h-100' {% endblock %}


{% block content %}
<div class="viewer-title">
    <div class="container viewer-title d-flex align-items-center justify-content-center text-center mt-4">
        <h4>
            <a class="brand px-3 py-3 col-12"
                href="{{ gse_metadata_single[species][geo_accession].get('gse_link', ['#'])[0]|safe }}" target="_blank"
                rel="noopener noreferrer" style="text-decoration: none">{{ geo_accession }}: {{
                    gse_metadata_single[species][geo_accession].get('title', ["Title Not Found"])[0]|safe}}
            </a>
        </h4>
    </div>
</div>




<div id="species" hidden>{{ species }}</div>
<div id="singlegse" hidden>{{ geo_accession }}</div>


<!-- Abstract small -->
<div class="row justify-content-center">
    <div class="col-10">

        <!-- Abstract large -->
        <div class="d-block ">
            <div class="h4 pl-2 mt-5 mb-2">Description (from GEO)</div>
            <div class="custom-col p-2">
                <p class="text-br-red font-weight-bold">Submission Date: <span class="font-weight-normal"
                        style="color:#212529">{{ (gse_metadata_single[species][geo_accession].get('submission_date')[0][:-5]
                        + ',' + gse_metadata_single[species][geo_accession].get('submission_date')[0][-5:])|safe }}</span>
                </p>
                {% for paragraph in gse_metadata_single[species][geo_accession].get('summary', ['Summary Not Found']) %}
                {% if loop.first %}
                <p class="text-br-red font-weight-bold">Summary: <span class="font-weight-normal"
                        style="color:#212529">{{ paragraph|safe }}</span></p>
                {% else %}
                <p>{{ paragraph|safe }}</p>
                {% endif %}
                {% endfor %}
                <p class="text-br-red font-weight-bold">GEO Accession ID: <a class="font-weight-normal"
                        style="color: #0e63c4"
                        href="{{ gse_metadata_single[species][geo_accession].get('gse_link', ['#'])[0]|safe }}" target="_blank"
                        rel="noopener noreferrer">{{
                            gse_metadata_single[species][geo_accession].get('geo_accession', ['GEO Accession ID Not Found'])[0]
                        }}</a></p>
                {% if gse_metadata_single[species][geo_accession].get('pubmed_id', ['No Pubmed ID'])[0]|safe != 'No Pubmed
                ID' %}
                <p class="text-br-red font-weight-bold">PMID: <a class="font-weight-normal" style="color: #0e63c4"
                        href="{{ gse_metadata_single[species][geo_accession].get('pubmed_link', ['#'])[0]|safe }}"
                        target="_blank" rel="noopener noreferrer">{{
                            gse_metadata_single[species][geo_accession].get('pubmed_id', ['No Pubmed ID'])[0]|safe }}</a></p>
                {% else %}
                <p class="text-br-red font-weight-bold">PMID: <span class="font-weight-normal"
                        style="color:#212529">N/A</span></p>
                {% endif %}
            </div>
        </div>
    </div>
</div>


<div class="row justify-content-center mb-3 mt-2">
    <div class="col-10">
        <div class="custom-col justify-content-center" style="overflow: visible">
            <div class="h4 pl-2 mt-4 mb-4 text-center"> Select Condition to compute Visualization and Analysis:</div>
            <div class="row justify-items-center justify-content-center mb-3 mt-3 text-center">
                <select name="Select Method" id="methodsingle" class="libpicker" >
                    {% for cond in study_conditions %}
                    <option value="{{cond}}">{{cond}}</option>
                    {% endfor %}
                </select>
            </div>
            <div class="h4 pl-2 mt-4 mb-4 text-center">PCA, tSNE, and UMAP Visualizations Labeled by Clusters:</div>
            <!-- <div class="row justify-content-center text-center">
                <button id="singleplots-button" class="btn btn-primary mx-auto mt-2 mb-3" style="width: 125px"
                    data-url-samp="{{ url_for('samples_api', geo_accession=geo_accession) }}"
                    data-gse-id="{{ gse_metadata_single[species][geo_accession].get('geo_accession', ['No GSE ID'])[0] }}"
                    data-gpl-id="{{ gse_metadata_single[species][geo_accession].get('cur_gpl', ['No GPL ID'])[0] }}">
                    Compute Visualizations
                </button>
            </div> -->
            <!-- <div id="singleplots-loading" class="m-3 mx-auto"></div> -->
            <div id="singleplots-loading" class="m-3"></div>
            <div class="row justify-content-center mt-4 singleplotclass centered">
                <ul class="nav nav-tabs">
                    <li class="nav-item">
                      <a class="nav-link active" data-toggle="tab" data-bs-target="#umap" href="#umap">UMAP</a>
                    </li>
                    <li class="nav-item">
                      <a class="nav-link" data-toggle="tab" data-bs-target="#tsne" href="#tsne">tSNE</a>
                    </li>
                    <li class="nav-item">
                      <a class="nav-link" data-toggle="tab" data-bs-target="#pca" href="#pca">PCA</a>
                    </li>
                  </ul>
                  
                  <!-- Tab panes -->
                <div class="h4 pl-2 mt-4 mb-4 text-center">
                  <div class="tab-content">
                    <div class="tab-pane container active" id="umap">
                        <div id="umap-plot" class="centered"></div>
                    </div>
                    <div class="tab-pane container fade" id="tsne">
                        <div id="tsne-plot" class="centered"></div>
                    </div>
                    <div class="tab-pane container fade" id="pca">
                        <div id="pca-plot" class="centered"></div>
                    </div>
                  </div>
                </div>
            </div>




            <div class="container px-4 pb-5" style="min-height: 100%;">

                <!-- Title -->
                <div class="row mt-4">
            
                    <div class="col-12">
                        <div class="h4 pl-2 mb-4">
                            {% if gse_metadata_single[species][geo_accession].get('type') == ['Expression profiling by array'] %}
                            Microarray Data Explorer
                            {% elif gse_metadata_single[species][geo_accession].get('type') == ['Expression profiling by high throughput
                            sequencing'] %}
                            RNA-Seq Data Explorer
                            {% else %}
                            Gene Expression Data Explorer
                            {% endif %}
                        </div>
                    </div>
                </div>
            
                <!-- Gene Expression Explorer -->
                <div class="row justify-content-center">
            
                    <!-- Boxplots -->
                    <div class="col-lg-8 col-xl-9 mb-4 justify-content-center">
            
                        <!-- Selector and plot -->
                        <div class="custom-col justify-content-center text-center">
                            Enter gene symbol:
                            <select id="gene-select" class="m-4 studysearch"
                                data-url-genes="{{ url_for('genes_api_single', geo_accession=geo_accession, condition=study_conditions[0]) }}"
                                data-geo-acc={{geo_accession}}></select>
                            <div id="boxplotloader" class="m-4"></div>
                            <div id="boxplot" class="plotly-plot chart m-4"
                                data-url-plot="{{ url_for('plot_api_single', geo_accession=geo_accession,condition=study_conditions[0]) }}"></div>
                        </div>
            

            
                        <div class="d-none d-lg-block">
            
                        </div>
            
                    </div>
            
                    <!-- Table -->
                    <div class="col-lg-4 col-xl-3 p-2">
                        <p class="m-3" style="font-size: 0.97rem;">Select conditions below to toggle them from the plot:</p>
            
                        <table id="condition-table" class="text-center">
                            <thead>
                                <tr>
    
                                    <th>Clusters</th>
                                    <th>Cell Samples</th>
                                </tr>
                            </thead>
                            <tbody id="singlecell">
                                {% for condition in metadata_dict %}

                                <tr>

                                    <td class="p-0 ml-3 border"><button
                                            style="background-color: #ECEFF1; color: black; width: max-content;"
                                            class="btn m-0 rounded-0 py-0 condition-btn active plotted " data-toggle="button"
                                             autocomplete="off"
                                            data-group_label="{{ condition }}">{{ condition|safe }} </button>
                                    </td>
            
                                    <td class="border" style="padding: 0px 18px; width: max-content;">
                                        <button
                                            style="width: max-content;"
                                            class="btn-custom btn-group-sm btn-collapse collapsed d-flex align-items-center text-center"
                                            data-toggle="collapse" data-target="#samples-{{ gse_id }}" aria-expanded="false"
                                            aria-controls="samples-{{ gse_id }}">
                                            <div class="text">{{ metadata_dict_samples[condition] }} Cells
                                            </div>
                                        </button>

            
                                    </td>
            
            
                                </tr>
                                {% endfor %}

                            </tbody>
                        </table>
                    </div>
                </div>
            </div>




        
            <div class="h4 pl-2 mt-4 mb-4 text-center">Differential Gene Expression Analysis</div>

            
            <div class="h6 text-center mt-3"> Select differential expression analysis method:</div>
            <div class="row justify-items-center justify-content-center mb-3 mt-3 text-center">
                <select name="Select Method" id="method" class="libpicker" style="width: 100px;">
                    <option value="limma">limma</option>
                    <option value="edgeR">edgeR</option>
                    <option value="DESeq2">DESeq2</option>
                </select>
            </div>
            <div id="normalization" style="display: none;">
                <div class="h6 text-center mb-2"> Select Normalization Methods:</div>
                    <div class="row justify-items-center justify-content-center">
                        <div class="h9 m-2">logCPM normalization:</div>
                        <label class="switch" for="logCPM">
                            <input type="checkbox" id="logCPM" checked>
                            <span class="slider round"></span>
                        </label>
                        <div class="h9 m-2">log normalization:</div>
                        <label class="switch" for="log"> 
                            <input type="checkbox" id="log">
                            <span class="slider round"></span>
                        </label>
                        <div class="h9 m-2">Z normalization:</div>
                        <label class="switch" for="z"> 
                            <input type="checkbox" id="z" checked>
                            <span class="slider round"></span>
                        </label>
                        <div class="h9 m-2">Quantile normalization:</div>
                        <label class="switch" for="q"> 
                            <input type="checkbox" id="q">
                            <span class="slider round"></span>
                        </label>
                    </div>
            </div>

            <div class="row justify-content-center">
                <button id="dge-button" class="btn btn-primary mx-auto mt-2 mb-3" style="width: 125px"
                    data-url-samp="{{ url_for('samples_api', geo_accession=geo_accession) }}"
                    data-gse-id="{{ gse_metadata_single[species][geo_accession].get('geo_accession', ['No GSE ID'])[0] }}"
                    data-gpl-id="{{ gse_metadata_single[species][geo_accession].get('cur_gpl', ['No GPL ID'])[0] }}">
                    Compute DGE
                </button>
            </div>

            <div id="dge-loading" class="m-3"></div>
            <div id="dge-plot" class="centered"></div>
            <div id="geneset-buttons" class="justify-content-center"></div>
            <div class="col-10 justify-content-center mx-auto" id="dge-table-area"></div>
    

            {% if 'Expression profiling by high throughput sequencing' == gse_metadata_single[species][geo_accession].get('type')[0] %}
            <div class="row justify-content-center mt-4">
                
                <div class="h4 pl-2 mt-4 mb-4 text-center">Single Cell RNA-seq Appyter</div>
                <div class="row text-center justify-content-center mt-2">
                    <div class="col-10">
                        <p> This pipeline enables you to analyze and visualize your bulk RNA sequencing datasets
                            with an array of downstream analysis and visualization tools. The pipeline includes:
                            PCA analysis, Clustergrammer interactive heatmap, library size analysis,
                            differential gene expression analysis, enrichment analysis, and L1000 small molecule
                            search.</p>
                        <img src="static/img/bulkdeg-thumbnail.png" alt="" class="img-fluid img-thumbnail">
                    </div>
                </div>
                

            </div>
            

            <div class="row justify-content-center">
                <button id="dgea-button" class="btn btn-primary mx-auto mt-2 mb-3"
                    data-url-samp="{{ url_for('samples_api', geo_accession=geo_accession) }}"
                    data-gse-id="{{ gse_metadata_single[species][geo_accession].get('geo_accession', ['No GSE ID'])[0] }}"
                    data-gpl-id="{{ gse_metadata_single[species][geo_accession].get('cur_gpl', ['No GPL ID'])[0] }}">
                    Open in <img src="{{ url_for('static', filename='img/appyters_logo.svg')}}" style="width: 100px;">
                </button>
                
            </div>
            {% endif %}

            <div id="dgea-loading" class="m-3"></div>

        </div>



    </div>

</div>






{% endblock %}

{% block footer %}
<script>
const el = document.getElementById('method');

const box = document.getElementById('normalization');

el.addEventListener('change', function handleChange(event) {
  if (event.target.value === 'DESeq2') {
    box.style.display = 'block';
  } else {
    box.style.display = 'none';
  }
});

$("#singleviewer").addClass('active'); 
$("#singleviewer").parents('li').addClass('active');


generate_single_plots()
const single_substudy = document.getElementById('methodsingle');

// single_substudy.addEventListener('change', function handleChange(event) {

//     var gse = document.getElementById("singlegse").innerText
//     var species = document.getElementById("species").innerText
//     var condition_group = document.getElementById("methodsingle").value
//     generate_single_plots()
//     var gsedata = JSON.stringify({'gse': gse, 'species': species, 'conditiongroup':condition_group});
//     console.log(gsedata)
//     console.log('In the change in viewer')
//     $.ajax({
//         url: "/getclusterdata",
//         contentType: 'application/json',
//         type: "POST",
//         data: gsedata,
//         dataType: 'json',
//     }).done(function(response) {

//         const classes = response['classes']
//         const metadict = response['metadict']
//         console.log(classes)
//         console.log(metadict)
//         document.getElementById("singlecell").innerHTML = ''
//         tabletext = ''
//         for (var k = 0; k < classes.length; k++) {
//                 tabletext += `<tr>
//                         <td class="p-0 ml-3 border"><button
//                                 style="background-color: #ECEFF1; color: black; width: max-content;"
//                                 class="btn m-0 rounded-0 py-0 condition-btn active plotted " data-toggle="button"
//                                 autocomplete="off"
//                                 data-group_label="${classes[k]}">${classes[k]}</button>
//                         </td>

//                         <td class="border" style="padding: 0px 18px; width: max-content;">
//                             <button
//                                 style="width: max-content;"
//                                 class="btn-custom btn-group-sm btn-collapse collapsed d-flex align-items-center text-center"
//                                 data-toggle="collapse" data-target="#samples-${gse}" aria-expanded="false"
//                                 aria-controls="samples-${gse}">
//                                 <div class="text">${metadict[classes[k]]} Cells
//                                 </div>
//                             </button>


//                         </td>
//                         </tr>`
//             }
//             document.getElementById("singlecell").innerHTML = tabletext


//     });

// });

</script>



{% endblock %}