{% extends 'base.html' %}

{% block body %}
<div class="viewer-title">
    <div class="container viewer-title d-flex align-items-center justify-content-center text-center mt-4">
        <h4>
            <a class="brand px-3 py-3 col-12"
                href="{{ gse_metadata[species][geo_accession].get('gse_link', ['#'])[0]|safe }}" target="_blank"
                rel="noopener noreferrer" style="text-decoration: none">{{ geo_accession }}: {{
                gse_metadata[species][geo_accession].get('title', ["Title Not Found"])[0]|safe}}
            </a>
        </h4>
    </div>
</div>

<div id='gse' style="display: none;">{{ geo_accession }}</div>
<div class="container px-4 pb-5" style="min-height: 100%;">

    <!-- Title -->
    <div class="row mt-4">

        <div class="col-12">
            <div class="h4 pl-2 mb-4 row">
                {% if gse_metadata[species][geo_accession].get('type') == ['Expression profiling by array'] %}
                Microarray Data Explorer
                <div class="tooltip ml-2">
                    <img src="../static/img/more_info.png" alt="Info" style="visibility: visible !important; opacity: 1 !important; width: 20px;"/>
                    <span class="tooltiptext">
                        Raw gene Expression data is sourced from GEO, and the appropriate db package for mapping probes to gene symbols was sourced from the <a href="https://www.bioconductor.org/packages/release/data/annotation/" target="_blank" rel="noopener noreferrer">Bioconductor AnnotationData packages</a>. 
                        You can read more about microarray data <a href="https://pubmed.ncbi.nlm.nih.gov/17660860/" target="_blank" rel="noopener noreferrer">here</a>.
                    </span>
                </div>
                {% elif gse_metadata[species][geo_accession].get('type') == ['Expression profiling by high throughput
                sequencing'] %}
                RNA-Seq Data Explorer
                <div class="tooltip ml-2">
                    <img src="../static/img/more_info.png" alt="Info" style="visibility: visible !important; opacity: 1 !important; width: 20px;"/>
                    <span class="tooltiptext">
                        Gene counts are sourced from ARCHS4, which provides uniform alignment of GEO samples. 
                        You can learn more about ARCHS4 and its pipeline <a href="https://maayanlab.cloud/archs4/" target="_blank" rel="noopener noreferrer">here</a>.
                    </span>
                </div>
                {% else %}
                Gene Expression Data Explorer
                <div class="tooltip ml-2">
                    <img src="../static/img/more_info.png" alt="Info" style="visibility: visible !important; opacity: 1 !important; width: 20px;"/>
                    <span class="tooltiptext">
                        Gene counts are sourced from ARCHS4, which provides uniform alignment of GEO samples. 
                        You can learn more about ARCHS4 and its pipeline <a href="https://maayanlab.cloud/archs4/" target="_blank" rel="noopener noreferrer">here</a>.
                    </span>
                </div>
                {% endif %}
            </div>
        </div>
    </div>

    

    <!-- Gene Expression Explorer -->
    <div class="row justify-content-center">
        

        <!-- Boxplots -->
        <div class="col-lg-8 col-xl-9 mb-4 justify-content-center">
            
            <!-- Selector and plot -->
            <div class="custom-col justify-content-center text-center">
                Enter gene symbol:
                <select id="gene-select" class="m-4 studysearch"
                    data-url-genes="{{ url_for('genes_api', geo_accession=geo_accession, species_name=species_name) }}"
                    data-geo-acc={{geo_accession}}></select>
                <div id="boxplot" class="plotly-plot chart m-4"
                    data-url-plot="{{ url_for('plot_api', geo_accession=geo_accession) }}"></div>
            </div>

            <!-- Abstract large -->
            <div class="d-none d-lg-block">
                <div class="h4 pl-2 mt-5 mb-4">Description</div>
                <div class="custom-col p-2">
                    <p class="text-br-red font-weight-bold">Submission Date: <span class="font-weight-normal"
                            style="color:#212529">{{
                            (gse_metadata[species][geo_accession].get('submission_date')[0][:-5]
                            + ',' + gse_metadata[species][geo_accession].get('submission_date')[0][-5:])|safe }}</span>
                    </p>
                    {% for paragraph in gse_metadata[species][geo_accession].get('summary', ['Summary Not Found']) %}
                    {% if loop.first %}
                    <p class="text-br-red font-weight-bold">Summary: <span class="font-weight-normal"
                            style="color:#212529">{{ paragraph|safe }}</span></p>
                    {% else %}
                    <p>{{ paragraph|safe }}</p>
                    {% endif %}
                    {% endfor %}
                    <p class="text-br-red font-weight-bold">GEO Accession ID: <a class="font-weight-normal"
                            style="color: #0e63c4"
                            href="{{ gse_metadata[species][geo_accession].get('gse_link', ['#'])[0]|safe }}"
                            target="_blank" rel="noopener noreferrer">{{
                            gse_metadata[species][geo_accession].get('geo_accession', ['GEO Accession ID Not Found'])[0]
                            }}</a></p>
                    {% if gse_metadata[species][geo_accession].get('pubmed_id', ['No Pubmed ID'])[0]|safe != 'No Pubmed
                    ID' %}
                    <p class="text-br-red font-weight-bold">PMID: <a class="font-weight-normal" style="color: #0e63c4"
                            href="{{ gse_metadata[species][geo_accession].get('pubmed_link', ['#'])[0]|safe }}"
                            target="_blank" rel="noopener noreferrer">{{
                            gse_metadata[species][geo_accession].get('pubmed_id', ['No Pubmed ID'])[0]|safe }}</a></p>
                    {% else %}
                    <p class="text-br-red font-weight-bold">PMID: <span class="font-weight-normal"
                            style="color:#212529">N/A</span></p>
                    {% endif %}
                </div>
            </div>

            <div class="d-none d-lg-block">

            </div>

        </div>

        <!-- Table -->
        <div class="col-lg-4 col-xl-3 p-2">
            <p class="m-3" style="font-size: 0.97rem;">Select conditions below to toggle them from the plot:</p>

            <table id="condition-table" class="text-center">
                <thead>
                    <tr>
                        <th>GROUP</th>
                        <th>CONDITION</th>
                        <th>SAMPLES</th>
                    </tr>
                </thead>
                <tbody>
                    {% for group, conditions in metadata_dict.items() %}
                    {% set grouploop = loop %}
                    {% for condition in conditions %}
                    <tr>
                        {% if loop.first %}
                        <td rowspan="{{ conditions|length }}" class="align-middle" style="width: auto">
                            {{ group|safe }}
                        </td>
                        {% endif %}
                        <!-- <td class="border p-0"><button style="background-color: #ECEFF1; color: black; width: max-content;" class="btn w-100 m-0 rounded-0 py-0 condition-btn{{' active plotted' if grouploop.first}}" data-toggle="button" aria-pressed="{{ grouploop.first|lower }}" autocomplete="off" data-group_label="{{ condition }}">{{ condition|safe }}</button></td> -->
                        <td class="p-0 ml-3 border"><button
                                style="background-color: #ECEFF1; color: black; width: max-content;"
                                class="btn m-0 rounded-0 py-0 condition-btn{{' active plotted ' if grouploop.first}}"
                                data-toggle="button" aria-pressed="{{ grouploop.first|lower }}" autocomplete="off"
                                data-group_label="{{ condition  + ' ' + group }}">{{ condition|safe }} </button>
                        </td>
                        <td class="border" style="padding: 0px 18px">
                            <button
                                class="btn-custom btn-group-sm btn-collapse collapsed d-flex align-items-center text-center"
                                data-toggle="collapse" data-target="#samples-{{ gse_id }}" aria-expanded="false"
                                aria-controls="samples-{{ gse_id }}">
                                <div class="text">Show Sample IDs ({{ metadata_dict_samples[group + ' ' +
                                    condition]['count'] }})
                                </div>
                            </button>
                            <div class="collapse" id="samples-{{ gse_id }}">
                                {{ " ".join(metadata_dict_samples[group + ' ' + condition]['samples']) }}
                            </div>
                        </td>
                    </tr>
                    {% endfor %}
                    {% endfor %}
                </tbody>
            </table>
        </div>
    </div>
</div>

<div id="species" hidden>{{ species }}</div>
<div id="gse" hidden>{{ geo_accession }}</div>


<!-- Abstract small -->
<div class="row justify-content-center">
    <div class="col-10">

        <!-- Abstract large -->
        <div class="d-block d-lg-none">
            <div class="h4 pl-2 mt-5 mb-2">Description</div>
            <div class="custom-col p-2">
                <p class="text-br-red font-weight-bold">Submission Date: <span class="font-weight-normal"
                        style="color:#212529">{{ (gse_metadata[species][geo_accession].get('submission_date')[0][:-5]
                        + ',' + gse_metadata[species][geo_accession].get('submission_date')[0][-5:])|safe }}</span>
                </p>
                {% for paragraph in gse_metadata[species][geo_accession].get('summary', ['Summary Not Found']) %}
                {% if loop.first %}
                <p class="text-br-red font-weight-bold">Summary: <span class="font-weight-normal"
                        style="color:#212529">{{ paragraph|safe }}</span></p>
                {% else %}
                <p>{{ paragraph|safe }}</p>
                {% endif %}
                {% endfor %}
                <p class="text-br-red font-weight-bold">GEO Accession ID: <a class="font-weight-normal"
                        style="color: #0e63c4"
                        href="{{ gse_metadata[species][geo_accession].get('gse_link', ['#'])[0]|safe }}" target="_blank"
                        rel="noopener noreferrer">{{
                        gse_metadata[species][geo_accession].get('geo_accession', ['GEO Accession ID Not Found'])[0]
                        }}</a></p>
                {% if gse_metadata[species][geo_accession].get('pubmed_id', ['No Pubmed ID'])[0]|safe != 'No Pubmed
                ID' %}
                <p class="text-br-red font-weight-bold">PMID: <a class="font-weight-normal" style="color: #0e63c4"
                        href="{{ gse_metadata[species][geo_accession].get('pubmed_link', ['#'])[0]|safe }}"
                        target="_blank" rel="noopener noreferrer">{{
                        gse_metadata[species][geo_accession].get('pubmed_id', ['No Pubmed ID'])[0]|safe }}</a></p>
                {% else %}
                <p class="text-br-red font-weight-bold">PMID: <span class="font-weight-normal"
                        style="color:#212529">N/A</span></p>
                {% endif %}
            </div>
        </div>
    </div>
</div>

<div class="row justify-content-center mb-5 mt-2">
    <div class="col-7">
        <div class="custom-col justify-content-center text-center" style="overflow: visible">

                <div class="pl-2 mt-4 mb-4" style="display: inline-flex;"><h4>Visualize Samples</h4> 
                    <div class="tooltip ml-2">
                        <img src="../static/img/more_info.png" alt="Info" style="visibility: visible !important; opacity: 1 !important; width: 20px;"/>
                        <span class="tooltiptext">
                            Visualizations are precomputed using the Python package <a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-017-1382-0" target="_blank" rel="noopener noreferrer">scanpy</a> on the top 5000 most variable genes.
                        </span>
                    </div>
                </div>
                

            

            <div class="row justify-content-center mt-4 singleplotclass centered">
                <ul class="nav nav-tabs">
                    <li class="nav-item">
                        <a class="nav-link active" data-toggle="tab" data-bs-target="#umap" href="#umap">UMAP</a>
                    </li>
                    <li class="nav-item">
                        <a class="nav-link" data-toggle="tab" data-bs-target="#tsne" href="#tsne">tSNE</a>
                    </li>
                    <li class="nav-item">
                        <a class="nav-link" data-toggle="tab" data-bs-target="#pca" href="#pca">PCA</a>
                    </li>
                </ul>

                <!-- Tab panes -->
                <div class="h4 pl-2 mt-4 mb-4 text-center">
                    <div class="tab-content">
                        <div class="tab-pane container active" id="umap">
                            <div id="umap-plot" class="centered"></div>
                        </div>
                        <div class="tab-pane container fade" id="tsne">
                            <div id="tsne-plot" class="centered"></div>
                        </div>
                        <div class="tab-pane container fade" id="pca">
                            <div id="pca-plot" class="centered"></div>
                        </div>
                    </div>
                </div>
                <div id="singleplots-loading" class="m-3"></div>
            </div>
            <div id='plots-loading' class="text-center"></div>
        </div>
    </div>
</div>


<div class="row justify-content-center mb-3 mt-2">
    <div class="col-7">
        <div class="custom-col justify-content-center" style="overflow: visible">
            <div class="pl-2 mt-4 mb-4 justify-content-center" style="display: flex;">
                <h4>Precomputed Differential Gene Expression</h4> 
                <div class="tooltip ml-2">
                    <img src="../static/img/more_info.png" alt="Info" style="width: 20px;"/>
                    <span class="tooltiptext">
                        Differential expression signatures are automatically computed using the <a href="https://pubmed.ncbi.nlm.nih.gov/25605792/" target="_blank" rel="noopener noreferrer">limma R package</a>.
                        More options for differential expression are available to compute below.
                    </span>
                </div>
            </div>
            <p class="text-br-red text-center font-weight-bold">Signatures:</p>
            <div class="row justify-content-center mt-4">
                {% if dge_precomputed|length > 0 %}
                <select id="precomputed-dge-conditions" class="mt-2 mb-2 mr-2 libpicker text-center"
                    style="text-overflow: ellipsis; width: 60%;">
                    <option value="">Select a signature to view limma differential expression</option>
                    {% for sig in dge_precomputed %}
                    <option value="{{ sig }}">{{ sig }}</option>
                    {% endfor %}
                </select>
                {% else %}
                <p>No precomputed signatures are currently available for this study. You can compute differential gene
                    expression on the fly below: </p>
                {% endif %}
            </div>
            <div id="precomputed-dge-res" class="m-3"></div>
            <div id="precomputed-dge-res-buttons" class="m-3"></div>
        </div>
    </div>
</div>



<div class="row justify-content-center mb-3 mt-2">
    <div class="col-7">
        <div class="custom-col justify-content-center" style="overflow: visible">
            <h4 class="text-center mt-3">Select conditions:</h4>

            <div class="row justify-content-center mt-4">
                <div class="w-25">
                    <p class="text-br-red text-center font-weight-bold">Control Condition</p>
                    <select id="condition-select-control" class="mt-2 mb-2 mr-2"
                        data-url-control="{{ url_for('conditions_api', geo_accession=geo_accession) }}"></select>
                </div>
                <div class="w-25">
                    <p class="text-br-red text-center font-weight-bold">Perturbation Condition</p>
                    <select id="condition-select-perturb" class="mt-2 mb-2 ml-2"
                        data-url-perturb="{{ url_for('conditions_api', geo_accession=geo_accession) }}"></select>
                </div>

            </div>
            <div class="row text-center">
                <p class="mx-auto"> Only conditions with at least 1 replicate are available to select</p>
            </div>

            <div class="h4 pl-2 mt-4 mb-4 justify-content-center" style="display: flex;">
                Differential Gene Expression Analysis
                <div class="tooltip ml-2">
                    <img src="../static/img/more_info.png" alt="Info" style="width: 20px;"/>
                    <span class="tooltiptext">
                        Differential expression signatures can be computed using <a href="https://pubmed.ncbi.nlm.nih.gov/24650281/" target="_blank" rel="noopener noreferrer">DESeq2</a> or <a href="https://pubmed.ncbi.nlm.nih.gov/24650281/" target="_blank" rel="noopener noreferrer">characteristic direction</a>.
                    </span>
                </div>
            </div>


            <div class="h6 text-center mt-3"> Select differential expression analysis method:</div>
            <div class="row justify-items-center justify-content-center mb-3 mt-3 text-center">
                <select name="Select Method" id="method" class="libpicker" style="width: auto;">
                    <option value="characteristic_direction">Characteristic Direction</option>
                    {% if 'Expression profiling by high throughput sequencing' ==
                    gse_metadata[species][geo_accession].get('type')[0] %}
                    <option value="DESeq2">DESeq2</option>
                    {% endif %}
                </select>
            </div>

            <div class="row justify-content-center">
                <button id="dge-button" class="btn btn-primary mx-auto mt-2 mb-3" style="width: 125px"
                    data-url-samp="{{ url_for('samples_api', geo_accession=geo_accession) }}"
                    data-gse-id="{{ gse_metadata[species][geo_accession].get('geo_accession', ['No GSE ID'])[0] }}"
                    data-gpl-id="{{ gse_metadata[species][geo_accession].get('cur_gpl', ['No GPL ID'])[0] }}">
                    Compute DGE
                </button>
            </div>

            <div id="dge-loading" class="m-3"></div>
            <div id="dge-plot" class="centered"></div>
            <div id="geneset-buttons" class="justify-content-center"></div>
            <div class="col-10 justify-content-center mx-auto" id="dge-table-area"></div>


            {% if 'Expression profiling by high throughput sequencing' ==
            gse_metadata[species][geo_accession].get('type')[0] %}
            <div class="row justify-content-center mt-4">

                <div class="h4 pl-2 mt-4 mb-4 text-center">Bulk RNA-seq Appyter</div>
                <div class="row text-center justify-content-center mt-2">
                    <div class="col-10">
                        <p> This pipeline enables you to analyze and visualize your bulk RNA sequencing datasets
                            with an array of downstream analysis and visualization tools. The pipeline includes:
                            PCA analysis, Clustergrammer interactive heatmap, library size analysis,
                            differential gene expression analysis, enrichment analysis, and L1000 small molecule
                            search.
                        </p>
                        <img src="static/img/bulkdeg-thumbnail.png" alt="" class="img-fluid img-thumbnail">
                    </div>
                </div>


            </div>


            <div class="row justify-content-center">
                <button id="dgea-button" class="btn btn-primary mx-auto mt-2 mb-3"
                    data-url-samp="{{ url_for('samples_api', geo_accession=geo_accession) }}"
                    data-gse-id="{{ gse_metadata[species][geo_accession].get('geo_accession', ['No GSE ID'])[0] }}"
                    data-gpl-id="{{ gse_metadata[species][geo_accession].get('cur_gpl', ['No GPL ID'])[0] }}">
                    Open in <img src="{{ url_for('static', filename='img/appyters_logo.svg')}}" style="width: 100px;">
                </button>

            </div>
            {% endif %}

            <div id="dgea-loading" class="m-3"></div>
        </div>
    </div>
</div>






{% endblock %}

{% block footer %}
<script>

    const el = document.getElementById('method');
    const box = document.getElementById('normalization');
    const gse = document.getElementById("gse").innerText
    const species = document.getElementById("species").innerText
    const loading = "<div class='loadingspinner'><div id='square1'></div><div id='square2'></div><div id='square3'></div><div id='square4'></div><div id='square5'></div></div>";

    const precomputed_sig = document.getElementById('precomputed-dge-conditions');
    if (precomputed_sig) {
        precomputed_sig.addEventListener("change", (event) => {
            document.getElementById('precomputed-dge-res').innerHTML = "<div class='text-center'>Retrieving results...</di>";
            if (precomputed_sig.value != '') {
                fill_precomputed_dge_table()
            } else {
                document.getElementById('precomputed-dge-res').innerHTML = "";
            }
        });
    }

    function fill_precomputed_dge_table() {
        const precomputed_res = document.getElementById('precomputed-dge-res')
        var gsedata = JSON.stringify({ 'sig': precomputed_sig.value, 'species': species });
        $.ajax({
            url: "api/precomputed_dge",
            contentType: 'application/json',
            type: "POST",
            dataType: 'json',
            data: gsedata,
        }).done(function (data) {

            var tabletext = `<table id='table-precomputed-dge' class='styled-table' style:'width=100%; vertical-align:top;'><thead><tr><th>Gene</th>`
            const genes = Object.keys(data)
            const columns = Object.keys(data[genes[0]])
            columns.forEach(cname => {
                tabletext += `<th>${cname}</th>`
            })
            tabletext += `</tr><tbody>`;

            const currURL = window.location.href.split('/')

            var url = currURL.join('/') + 'singlegene'
            for (var k = 0; k < genes.length; k++) {
                tabletext += `<tr><td><a href='${url}' onclick="setGene('${genes[k]}')" target='_blank'>${genes[k]}<a/></td>`
                columns.forEach(cname => {
                    tabletext += "<td>" + data[genes[k]][cname] + "</td>"
                })
                tabletext += "</tr>"
            }
            tabletext += "</tbody></table>";
            precomputed_res.innerHTML = tabletext

            var table = $(`#table-precomputed-dge`).DataTable({
                order: [[5, 'asc']],
                dom: 'Bfrtip',
                buttons: [
                    'copy', { extend: 'csv', title: precomputed_sig.value }
                ]
            })

            var adjpvals = table.column(5).data()
            var pvals = table.column(4).data()
            var logfc = table.column(1).data()
            var genes_list = table.column(0).data()


            genes_list = genes_list.map(x => x.replace(/<\/?[^>]+(>|$)/g, ""))

            var genelist_buttons =
            `<div class="row justify-content-center mx-auto text-center">
                <div class="h7">Submit the top</div>
                <input class="" id='numgenes' type='number' step='1' value='100' pattern='[0-9]' min='1' class='m-2'
                    style='width: 50px; height: 30px; margin-left: 10px; margin-right: 10px;' />
                <select id='dir' style='margin-left: 10px; margin-right: 10px; height: fit-content;'>
                    <option value='up'>upregulated</option>
                    <option value='down'>downregulated</option>
                </select>
                <div class="h7"> differentially expressed genes with a
                    <select id='col-to-use' style='margin-left: 10px; margin-right: 10px; height: fit-content;'>
                        <option value='pval'>p-value</option>
                        <option value='adjpval'>adjusted p-value</option>
                    </select>
                </div>
                <div class=" h7"> less than </div>
                <input class='' id='signifigance' type='number' step='.01' value='.05' max='1'
                    style='width: 50px; height: 30px; margin-left: 10px; margin-right: 10px;"' />
                <div class="h7"> to</div>
            </div>
            <div class="row justify-content-center mx-auto text-center">
                <button class="btn btn-primary btn-group-sm m-2"
                    onclick="submit_geneset('${genes_list.join(',')}', '${adjpvals.join(',')}', '${pvals.join(',')}', '${logfc.join(',')}')">Gene Set Queries</button>
            </div>
            <div class="row justify-content-center mx-auto text-center">
                <button type="button" class="btn btn-primary btn-group-sm m-2"
                    onclick="filter_and_submit_to_enrichr('${genes_list.join(',')}', '${adjpvals.join(',')}', '${pvals.join(',')}', '${logfc.join(',')}', '${precomputed_sig.value}')">
                    Enrichr
                    <img src="/static/img/enrichrlogo.png" class="img-fluid mr-3" style="width: 45px" alt="Enrichr">
                </button>
                <button type="button" class="btn btn-primary btn-group-sm m-2"
                    onclick="filter_and_submit_to_kg('${genes_list.join(',')}', '${adjpvals.join(',')}', '${pvals.join(',')}', '${logfc.join(',')}', '${precomputed_sig.value}')">
                    Enrichr-KG
                    <img src="/static/img/enrichr-kg.png" class="img-fluid mr-3" style="width: 45px" alt="Enrichr">
                </button>
            </div>`
            document.getElementById('precomputed-dge-res-buttons').innerHTML = genelist_buttons;
            
            return
        })
    }

    function generate_plots() {
        document.getElementById("umap-plot").innerHTML = "";
        document.getElementById("tsne-plot").innerHTML = "";
        document.getElementById("pca-plot").innerHTML = "";

        var gsedata = JSON.stringify({ 'gse': gse, 'species': species });
        $.ajax({
            url: "api/bulksampvis",
            contentType: 'application/json',
            type: "POST",
            dataType: 'json',
            data: gsedata,
        }).done(function (response) {
            var umap_plot = response['umapplot']
            var pca_plot = response['pcaplot']
            var tsne_plot = response['tsneplot']

            if (pca_plot == 'None') {
                document.getElementById('umap').innerText = 'No visualization avaiable'
                document.getElementById('tsne').innerText = 'No visualization avaiable'
                document.getElementById('pca').innerText = 'No visualization avaiable'
            } else {
                window.Bokeh.embed.embed_item(umap_plot)
                window.Bokeh.embed.embed_item(pca_plot)
                window.Bokeh.embed.embed_item(tsne_plot)
            }
        });
    }

    generate_plots()


    $("#viewer").addClass('active');
    $("#viewer").parents('li').addClass('active');

</script>
{% endblock %}